src/charm.py

```diff
@@ -64,6 +66,87 @@ def evaluate(self, csr: bytes, relation_id: int, requirer_csrs: list[RequirerCSR
         return True
 
 
+class AllowedFields:
+    """Filter the CSR so as to only allow CSRs that match the given regexes for the CSR fields."""
+
+    COMMON_NAME_OID = x509.OID_COMMON_NAME
+    ORGANIZATION_OID = x509.OID_ORGANIZATION_NAME
+    EMAIL_OID = x509.OID_EMAIL_ADDRESS
+    COUNTRY_CODE_OID = x509.OID_COUNTRY_NAME
+
+    def __init__(self, filters: dict):
+        self.field_filters = filters
+
+    def _evaluate_subject(
+        self,
+        challenge: str,
+        subject: x509.Name,
+        oid: x509.ObjectIdentifier,
+        field_optional: bool = False,
+    ) -> str:
+        pattern = re.compile(challenge)
+        name_attributes = subject.get_attributes_for_oid(oid)
+        if not field_optional and not name_attributes:
+            return "field not found"
+        if any(not pattern.match(str(val.value)) for val in name_attributes):
+            return "field validation failed"
+        return ""
+
+    def _evaluate_sans(
+        self, challenge: str, san: x509.SubjectAlternativeName, type: Literal["dns", "ip", "oid"]
```

Comment:
Why are we using literals instead of something typed like an enum? or even better, the `x509` class itself?

It seems like we're introducing room for error here and making discoverability more difficult.

nit: I also think we might be overloading this function too much, and leaking Go-style into this code. This could easily be 3 separate functions that only need to return a bool. The shared part is minimal, and we're actually complicates the logic a bit. Not a big deal right now, but if this expands in scope, I think it gets messier. I'd go with 3 functions like the following.

```
    def _validate_sans_dns(self, challenge: str, san: x509.SubjectAlternativeName) -> bool:
        pattern = re.compile(challenge)
        return any(not pattern.match(domain) for domain in san.get_values_for_type(x509.DNSName))
```

Reply:
An enum already uses literals as type hints internally, so if they are not reused elsewhere I tend to avoid them. What error and what discoverability would be missing with them?

Also, the reason I didn't do `x509.DNSName | x509.IPAddress | x509.RegisteredID` was because the type hint for `san.get_values_for_type` is a bit borked and broke the static type checks. Fixing those type checks the "correct" way made the function look complex.

I don't think we're overloading this function (and also don't see how a match statement is Go-style), but loading it appropriately. It's clear what is being done and how, and adding 3 functions on the class level instead of here just moves the complexity into the class itself. If we expand, we only have to add 2 lines here, only within the scope of a function where it would be relevant, instead of another function and growing the list of functions for `self` in every other function that will definitely not need this information (something we tend to do a lot). I think it would be much more messy if the signature for this filter class looks like:

```
evaluate
_evaluate_subject
_evaluate_sans_...
_evaluate_sans_...
_evaluate_sans_...
_evaluate_sans_...
_evaluate_sans_...
```

Reply:
> What error and what discoverability would be missing without them?

error: I guess the main thing would be that this relies on the type checker, and not the compiler. Maybe that's not a good argument, but since type checking is relatively new I'm a little anxious about it.

discoverability: searching for an enum usage, renaming, etc... objectively easier with an enum.

> the type hint for san.get_values_for_type is a bit borked and broke the static type checks. Fixing those type checks the "correct" way made the function look complex.

This seems really odd, but I don't know what you were seeing. I can't see any reason why it wouldn't work, but I guess this feeds into my mistrust of the type checker.

> adding 3 functions on the class level instead of here just moves the complexity into the class itself

We can agree to disagree here, but from my perspective we've combined 3 functions into one, and are just using a parameter to switch on them. It already seems unnecessarily complex since I need to read the whole thing to follow what will happen in the case I'm interested in, rather than the 2 lines it would be independently. Might seem reasonable now as it's <10 lines, but these things tend to get messy as they expand.

>  I think it would be much more messy if the signature for this filter class looks like

I'd rather have lots of small functions than one big one with multiple paths.

> also don't see how a match statement is Go-style

I'm referring to returning strings and empty strings instead of a simple bool. It seems like we're using the return value like a Go-style `err` when we could just use a simple true/false to define if the pattern matched or not.


Anyway, this is all pretty minor, do with it what you want. It's mostly just opinion.

Reply:
> error: I guess the main thing would be that this relies on the type checker, and not the compiler. Maybe that's not a good argument, but since type checking is relatively new I'm a little anxious about it.

There is no compiler in Python. An enum is syntactic sugar for a Literal type, which you can see if you hover over the variable that's typed with an enum. You can pass in any variable to an enum typed parameter which will pass the linter that allows python to run in an interpolated way, just like how you can pass in any value to a Literal type. The static checker will do the same thing either way.

with discoverability, I'd use an enum if there was any need for this elsewhere, but I would avoid prematurely class-ifying by creating an enum

> This seems really odd, but I don't know what you were seeing. I can't see any reason why it wouldn't work, but I guess this feeds into my mistrust of the type checker.

Because it's marked as type[x509.DNSName] instead of x509.DNSName, extracting it without the type checker complaining is very weird. If you can do it I'd be impressed.

> but from my perspective we've combined 3 functions into one, and are just using a parameter to switch on them

I think that's probably why I write functions in the first place. What's the point of writing a function if you're not using parameters to branch out. Otherwise the only point would be the assembly style GOTO functionality. We are not switching for things that are irrelevant and coupling irrelevant functionality into one, we are handling different cases for the same function, just like the built in `get_values_for_type()` function.

> I'd rather have lots of small functions than one big one with multiple paths.

I'd much rather have to reason about a single function argument than try to pick the function needle in the class haystack

> I'm referring to returning strings and empty strings instead of a simple bool. It seems like we're using the return value like a Go-style err when we could just use a simple true/false to define if the pattern matched or not.

ah i see, that's fair. That could probably be refactored too.

> Anyway, this is all pretty minor, do with it what you want. It's mostly just opinion.

I like arguing with you :D

Reply:
> Because it's marked as type[x509.DNSName] instead of x509.DNSName, extracting it without the type checker complaining is very weird. If you can do it I'd be impressed.

Oh I see, because those are classes and it just wants the type. Yeah, odd. I see what you mean now.

> I like arguing with you :D

That's just because I'm an easy target. Go pick on someone on your level like @ghislainbourgeois .

Reply:
Oh, but if you still want me to impress you, you can just do `type[x509.DNSName | x509.IPAddress | x509.RegisteredID]'

Reply:
> Oh, but if you still want me to impress you, you can just do `type[x509.DNSName | x509.IPAddress | x509.RegisteredID]`

maybe in a different PR :D